src/pke/lib/keyswitch/keyswitch-hybrid.cpp

```diff
@@ -250,36 +262,48 @@ EvalKey<DCRTPoly> KeySwitchHYBRID::KeySwitchGen(const PrivateKey<DCRTPoly> oldKe
         DCRTPoly a(paramsQP, Format::EVALUATION, true);
         DCRTPoly b(paramsQP, Format::EVALUATION, true);
 
-        // The part with basis Q
-        for (usint i = 0; i < sizeQ; i++) {
-            const NativeInteger& qi = paramsQ->GetParams()[i]->GetModulus();
-            auto e0i                = e0.GetElementAtIndex(i);
-            auto e1i                = e1.GetElementAtIndex(i);
+        usint startPartIdx = numPerPartQ * part;
+        usint endPartIdx   = (sizeQ > startPartIdx + numPerPartQ) ? (startPartIdx + numPerPartQ) : sizeQ;
+
+        for (usint i = 0; i < startPartIdx; i++) {
+            auto e0i = e0.GetElementAtIndex(i);
+            auto e1i = e1.GetElementAtIndex(i);
+
+            auto ui = u.GetElementAtIndex(i);
+
+            auto newp0i = newp0.GetElementAtIndex(i);
+            auto newp1i = newp1.GetElementAtIndex(i);
+
+            a.SetElementAtIndex(i, newp1i * ui + ns * e1i);
+            b.SetElementAtIndex(i, newp0i * ui + ns * e0i);
+        }
+
+        for (usint i = startPartIdx; i < endPartIdx; i++) {
```

Comment:
I suggest combining three loops into a single loop. They appear to be the same except where we deal with the actual digit.

Reply:
updated